polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisAuthorizerImpl.java

```diff
@@ -602,15 +606,25 @@ public boolean hasTransitivePrivilege(
       Set<Long> activatedGranteeIds,
       PolarisPrivilege desiredPrivilege,
       PolarisResolvedPathWrapper resolvedPath) {
+    PolarisGrantManager grantManager =
+        grantManagerFactory.getGrantManagerForRealm(
+            CallContext.getCurrentContext().getRealmContext());
+    PolarisCallContext callContext = CallContext.getCurrentContext().getPolarisCallContext();
 
     // Iterate starting at the parent, since the most common case should be to manage grants as
     // high up in the resource hierarchy as possible, so we expect earlier termination.
-    for (ResolvedPolarisEntity resolvedSecurableEntity : resolvedPath.getResolvedFullPath()) {
-      Preconditions.checkState(
-          resolvedSecurableEntity.getGrantRecordsAsSecurable() != null,
-          "Got null grantRecordsAsSecurable for resolvedSecurableEntity %s",
-          resolvedSecurableEntity);
-      for (PolarisGrantRecord grantRecord : resolvedSecurableEntity.getGrantRecordsAsSecurable()) {
+    for (PolarisEntity resolvedSecurableEntity : resolvedPath.getRawFullPath()) {
+      PolarisGrantManager.LoadGrantsResult grantsResult =
+          grantManager.loadGrantsOnSecurable(
```

Comment:
I'm generally fine with this change, but I'd like to clarify one point. My impression from current code on `main` is that we want to run authZ check on the exact same state of Polaris data as what the API services use for generating API responses.

Is that a correct assumption?

If I am mistaken, that is fine and this change is fine too. However, if my assumption is valid, how do we make sure the loaded grant records match the other data, which is loaded independently by API services?

Reply:
If this is a reference to the same `PolarisMetaStoreManagerImpl.loadCachedEntryById` and `PolarisMetaStoreManagerImpl.loadGrantsOnSecurable` API difference you call out below, I do think there's more work to line the cache up with the metastore API calls. Ultimately, I think the `PolarisServerImpl` should be fetching the same cached grant records that are being returned to the Authorizer so that the return values and the authz assertions are the same.

Reply:
This thread is about the overall assumptions on the consistency of data in storage in the presence of concurrent changes.

With this code, grants for different objects in the same logical API-level operation may be loaded from different states of the data (e.g. results of two of more RDBMS transaction). So my question is whether this is a concern. I'm not proposing any particular approach, just wondering for my own understanding.